### PR TITLE
Multiple eth clients

### PR DESCRIPTION
**Pull request by tailingchen**
This PR implements the eth client that can connect to multiple eth targets.
This multiclient only implements some APIs of original ethclient.
#### Get-Type API
`multiclient` sends request to all ethclients, then waits until get `first` successful response and cancel other requests. Return error if all requests are fail.
#### Post-Type API
`multiclient` sends request to all ethclients, then waits for `all` responses. Return error if all requests are fail.

### CODE REVIEW

File: multiclient/client.go [hunk 1]
```diff
@@ -0,0 +1,531 @@
...
+
+	for _, opt := range opts {
+		if err := opt(mc); err != nil {
+			return nil, err
+		}
+	}
+	if len(mc.ethURLs) == 0 {
+		return nil, ErrNoEthClient
+	}
+	log.Info("Create Client", "urls", mc.ethURLs)
```
Comment (markya0616) [inline multiclient/client.go]: Suggest to remove it

File: multiclient/client.go [hunk 2]
```diff
@@ -0,0 +1,531 @@
...
+	mc.rpcClients = make([]*rpc.Client, len(mc.ethURLs))
+	for i, rawURL := range mc.ethURLs {
+		ctx, cancel := context.WithTimeout(context.Background(), dialTimeout)
```
Comment (markya0616) [inline multiclient/client.go]: How about pass an external context in?

File: multiclient/option.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	if err != nil {
+		log.Error("Failed to get service from consul", "serviceID", serviceID, "err", err)
+		return nil, err
+	}
+
+	ethURLs := make([]string, len(list))
+	for i, srv := range list {
+		addr := srv.ServiceAddress
+		if addr == "" {
+			addr = srv.Address
```
Comment (markya0616) [inline multiclient/option.go:75]: Just curious about what the difference between `srv.ServiceAddress` and `srv.Address` is

File: multiclient/client.go [hunk 4]
```diff
@@ -0,0 +1,531 @@
...
+
+func (mc *Client) CallContext(ctx context.Context, result interface{}, method string, args ...interface{}) error {
+	fns := make([]postFn, len(mc.rpcClients))
+	for i, c := range mc.rpcClients {
+		fns[i] = func(ctx context.Context) error {
+			return c.CallContext(ctx, result, method, args...)
+		}
+	}
+
+	err := postToAll(ctx, fns)
```
Comment (markya0616) [inline multiclient/client.go]: It should be `getFromAny` if one of the rpc is successful?

File: multiclient/client.go [hunk 5]
```diff
@@ -0,0 +1,531 @@
...
+
+func (mc *Client) BatchCallContext(ctx context.Context, b []rpc.BatchElem) error {
+	fns := make([]postFn, len(mc.rpcClients))
+	for i, c := range mc.rpcClients {
+		fns[i] = func(ctx context.Context) error {
+			return c.BatchCallContext(ctx, b)
+		}
+	}
+
+	err := postToAll(ctx, fns)
```
Comment (markya0616) [inline multiclient/client.go]: ditto

File: multiclient/option.go (hunk 3 above)
Comment (tailingchen) [reply multiclient/option.go:75]: `Address` is the IP address of the Consul node on which the service is registered.
`ServiceAddress` is the IP address of the service host — if empty, node address should be used.
https://www.consul.io/api/catalog.html#sample-response-3

**Review by yurenju — APPROVED** [review]

File: multiclient/client.go (hunk 1 above)
Comment (tailingchen) [reply multiclient/client.go]: Change log level to `Debug`.
Updated in `ed00a9f`.

File: multiclient/client.go (hunk 2 above)
Comment (tailingchen) [reply multiclient/client.go]: Updated in `ed00a9f`.

File: multiclient/client.go (hunk 4 above)
Comment (tailingchen) [reply multiclient/client.go]: Add `isPostToAll` param to decide to perform `getFromAny` or `postToAll`.
Updated in `26ccc64`.

**Review by markya0616 — APPROVED** [review]

**Pull request merged**